plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/client/ElasticsearchClient.java

```diff
@@ -538,45 +623,166 @@ public boolean indexExists(String index)
+                .map(field -> (IndexMetadata.ObjectType) field.getType())
+                .reduce((nestedSchema1, nestedSchema2) -> merge(index, prefix, nestedSchema1, nestedSchema2))
+                .get();
+        IndexMetadata.Field indexField = fields.iterator().next();
+        return new IndexMetadata.Field(indexField.isArray(), indexField.getName(), type);
+    }
+
+    private Set<IndexMetadata.Field> checkAndCorrectMismatchFields(String index, String parent, Set<IndexMetadata.Field> fields)
+    {
+        final Map<String, List<IndexMetadata.Field>> fieldsByName = fields.stream().collect(groupingBy(IndexMetadata.Field::getName));
+
+        final Set<IndexMetadata.Field> resultFields = fieldsByName.values().stream()
+                .filter(fieldList -> fieldList.size() == 1)
+                .map(fieldList -> fieldList.get(0))
+                .collect(toSet());
+
+        final Map<String, List<IndexMetadata.Type>> columnsWithMultipleDataTypes = fieldsByName.entrySet()
+                .stream()
+                .filter(fieldNameToFieldsList -> fieldNameToFieldsList.getValue().size() > 1)
+                .collect(toImmutableMap(Map.Entry::getKey, columnToFields -> columnToFields.getValue().stream()
+                        .map(IndexMetadata.Field::getType)
+                        .collect(toList())));
+
+        if (!columnsWithMultipleDataTypes.isEmpty() && failOnAliasSchemaMismatch) {
+            Set<String> failedColumns = columnsWithMultipleDataTypes.keySet();
+            if (!isNullOrEmpty(parent)) {
+                failedColumns = failedColumns.stream()
+                        .map(name -> getFlattenedKey(parent, name))
+                        .collect(toImmutableSet());
+            }
+            throw new TrinoException(ELASTICSEARCH_INVALID_METADATA, format("Table(Alias index) '%s' has columns '%s' having data type mismatch", index, failedColumns));
+        }
+
+        resultFields.addAll(correctColumnsWithMultipleDataTypes(columnsWithMultipleDataTypes));
+
+        return resultFields;
+    }
+
+    private Set<IndexMetadata.Field> correctColumnsWithMultipleDataTypes(Map<String, List<IndexMetadata.Type>> fieldsWithMultipleDataTypes)
+    {
+        return fieldsWithMultipleDataTypes.entrySet().stream()
+                .map(fieldType -> {
+                    // column with both PrimitiveType and non PrimitiveType casts as raw json
+                    final Class<? extends IndexMetadata.Type> firstTypeClazz = fieldType.getValue().get(0).getClass();
+                    final boolean isFieldPrimitive = fieldType.getValue().stream().allMatch(type -> type.getClass().equals(firstTypeClazz));
+
+                    String resultDataType = "text";
+                    if (isFieldPrimitive) {
+                        final Set<String> dataTypes = fieldType.getValue().stream()
+                                .map(type -> ((IndexMetadata.PrimitiveType) type).getName())
```

Comment:
would be good to have a check to make sure everything is primitive

Reply:
Its already checked @ line 774. If all the fields are primitive only, we enter this block.

Reply:
I'm a bit confused, could you help understand how the following always gives Primitive type?

```
final Class<? extends IndexMetadata.Type> firstTypeClazz = fieldType.getValue().get(0).getClass();
```

Reply:
We have to ensure that mismatch should be within primitive types. Mismatch cases are as below:

- **All FieldTypes are PrimitiveType**: Below check handles it by ensuring that all fieldTypes are of same type. 
- **One FieldType is PrimitiveType & Other is ObjectType/DataTimeType**: Below check ensures that `isFieldPrimitive` becomes `false` for such cases.
- **Both FieldTypes are ObjectType**: We never enter here in such cases. Because we intend to merge the nested fieldTypes inside the ObjectType and that is already handled.

https://github.com/sumannewton/trino/blob/7fee524a29fffb6a4e0565816b4dad7365b9bfcf/plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/client/ElasticsearchClient.java#L770-L771

Reply:
@phd3 Thoughts?